[Nested tensor]avoid using shape in python subclass NT, use _size instead

YuqingJ opened the pull request:
Summary:
calling tensor.shape will call torch_dispatch which adds more overhead.

Testing overhead difference in "NT + NT" operation:
**Before:**
the add operation takes ~300us
{F1167963824} 
**After:**
the add operation takes ~200us
 {F1167964056}

Test Plan: unit tests in test_nestedtensor

Differential Revision: D51949135

soulitzer reviewed (approved):
Nice!

jbschlosser commented on torch/nested/_internal/ops.py:
```diff
@@ -201,20 +201,18 @@ def jagged_binary_pointwise(func, *args, **kwargs):
-        raise RuntimeError(mismatch_error_msg)
-
+        raise RuntimeError(mismatch_error_msg.format(func.__name__, a._size, b._size))
+        # raise RuntimeError(mismatch_error_msg)
```
nit: delete commented-out code :)

jbschlosser reviewed (approved):
Good call, thanks!

YuqingJ committed:
[Nested tensor]avoid using shape in python subclass NT, use _size instead (#115371)

Summary:

calling tensor.shape will call torch_dispatch which adds more overhead.

Testing overhead difference in "NT + NT" operation:
**Before:**
the add operation takes ~300us
{F1167963824} 
**After:**
the add operation takes ~200us
 {F1167964056}

Test Plan: unit tests in test_nestedtensor

Reviewed By: soulitzer

Differential Revision: D51949135

YuqingJ commented:
@pytorchbot merge
